File: eggplant-pattern-extractor/src/ir.rs
use serde::Serialize;

#[derive(Debug, Clone, Copy, Serialize, PartialEq, Eq)]
pub struct TextSpan {
    pub start: usize,
    pub end: usize,
}

impl TextSpan {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ScopeKind {
    AddRuleCall,
    PatternFunction,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct ScopeInfo {
    pub kind: ScopeKind,
    pub text_range: TextSpan,
    pub pattern_range: Option<TextSpan>,
    pub action_range: Option<TextSpan>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NodeKind {
    Query,
    QueryLeaf,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct PatternNode {
    pub id: String,
    pub kind: NodeKind,
    pub dsl_type: String,
    pub label: String,
    pub range: TextSpan,
    pub inputs: Vec<String>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct PatternEdge {
    pub from: String,
    pub to: String,
    pub kind: String,
    pub index: usize,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct PatternConstraint {
    pub id: String,
    pub source_text: String,
    pub resolved_text: String,
    pub semantic_text: Option<String>,
    pub referenced_vars: Vec<String>,
    pub range: TextSpan,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct ActionEffect {
    pub id: String,
    pub effect_id: String,
    pub bound_var: Option<String>,
    pub source_text: String,
    pub semantic_text: Option<String>,
    pub referenced_pat_vars: Vec<String>,
    pub referenced_action_vars: Vec<String>,
    pub range: TextSpan,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct SeedFact {
    pub id: String,
    pub source_text: String,
    pub committed_root: String,
    pub referenced_vars: Vec<String>,
    pub range: TextSpan,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct DisplayTemplate {
    pub variant_name: String,
    pub template: String,
    pub fields: Vec<String>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct TypstTemplate {
    pub variant_name: String,
    pub template: String,
    pub fields: Vec<String>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct PrecedenceTemplate {
    pub variant_name: String,
    pub precedence: u16,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Severity {
    Error,
    Warning,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct Diagnostic {
    pub severity: Severity,
    pub message: String,
    pub range: Option<TextSpan>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct MathViewEntry {
    pub target_id: String,
    pub label: String,
    pub plain_source: String,
    pub colored_source: String,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct MathViewFormulaSource {
    pub plain: String,
    pub colored: String,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum MathViewConclusion {
    Rewrite {
        id: String,
        from: MathViewEntry,
        to: MathViewEntry,
    },
    Derive {
        id: String,
        entry: MathViewEntry,
    },
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct MathView {
    pub rule_name: String,
    pub premises: Vec<MathViewEntry>,
    pub side_conditions: Vec<String>,
    pub derivations: Vec<MathViewEntry>,
    pub conclusions: Vec<MathViewConclusion>,
    pub formula_source: MathViewFormulaSource,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct PatternIr {
    pub scope: ScopeInfo,
    pub nodes: Vec<PatternNode>,
    pub edges: Vec<PatternEdge>,
    pub roots: Vec<String>,
    pub constraints: Vec<PatternConstraint>,
    pub action_effects: Vec<ActionEffect>,
    pub seed_facts: Vec<SeedFact>,
    pub display_templates: Vec<DisplayTemplate>,
    pub typst_templates: Vec<TypstTemplate>,
    pub precedence_templates: Vec<PrecedenceTemplate>,
    pub diagnostics: Vec<Diagnostic>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub math_view: Option<MathView>,
}

File: eggplant-pattern-vscode/src/ir.ts
export interface TextSpan {
  start: number;
  end: number;
}

export interface ScopeInfo {
  kind: "add_rule_call" | "pattern_function";
  text_range: TextSpan;
  pattern_range: TextSpan | null;
  action_range: TextSpan | null;
}

export interface PatternNode {
  id: string;
  kind: "query" | "query_leaf";
  dsl_type: string;
  label: string;
  range: TextSpan;
  inputs: string[];
}

export interface PatternEdge {
  from: string;
  to: string;
  kind: string;
  index: number;
}

export interface PatternConstraint {
  id: string;
  source_text: string;
  resolved_text: string;
  semantic_text?: string | null;
  referenced_vars: string[];
  range: TextSpan;
}

export interface ActionEffect {
  id: string;
  effect_id: string;
  bound_var: string | null;
  source_text: string;
  semantic_text?: string | null;
  referenced_pat_vars: string[];
  referenced_action_vars: string[];
  range: TextSpan;
}

export interface SeedFact {
  id: string;
  source_text: string;
  committed_root: string;
  referenced_vars: string[];
  range: TextSpan;
}

export interface DisplayTemplate {
  variant_name: string;
  template: string;
  fields: string[];
}

export interface TypstTemplate {
  variant_name: string;
  template: string;
  fields: string[];
}

export interface PrecedenceTemplate {
  variant_name: string;
  precedence: number;
}

export interface Diagnostic {
  severity: "error" | "warning";
  message: string;
  range: TextSpan | null;
}

export interface MathViewEntry {
  target_id: string;
  label: string;
  plain_source: string;
  colored_source: string;
}

export type MathViewConclusion =
  | {
      kind: "rewrite";
      id: string;
      from: MathViewEntry;
      to: MathViewEntry;
    }
  | {
      kind: "derive";
      id: string;
      entry: MathViewEntry;
    };

export interface MathViewFormulaSource {
  plain: string;
  colored: string;
}

export interface MathViewModel {
  rule_name: string;
  premises: MathViewEntry[];
  side_conditions: string[];
  derivations: MathViewEntry[];
  conclusions: MathViewConclusion[];
  formula_source: MathViewFormulaSource;
}

export interface PatternIr {
  scope: ScopeInfo;
  nodes: PatternNode[];
  edges: PatternEdge[];
  roots: string[];
  constraints: PatternConstraint[];
  action_effects: ActionEffect[];
  seed_facts: SeedFact[];
  display_templates: DisplayTemplate[];
  typst_templates: TypstTemplate[];
  precedence_templates: PrecedenceTemplate[];
  diagnostics: Diagnostic[];
  math_view?: MathViewModel | null;
}
